IdentityCore/src/cache/accessor/MSIDSharedTokenCache.m

```diff
@@ -0,0 +1,238 @@
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDSharedTokenCache.h"
+#import "MSIDTokenCacheKey.h"
+#import "MSIDTelemetryEventStrings.h"
+#import "MSIDTelemetry+Internal.h"
+#import "MSIDAccount.h"
+#import "MSIDAdfsToken.h"
+
+@interface MSIDSharedTokenCache()
+{
+    // Primary cache format
+    id<MSIDSharedCacheFormat> _primaryFormat;
+    
+    // All shared formats starting with the primary
+    NSArray<id<MSIDSharedCacheFormat>> *_allFormats;
+}
+
+@end
+
+@implementation MSIDSharedTokenCache
+
+#pragma mark - Init
+
+- (instancetype)initWithPrimaryCacheFormat:(id<MSIDSharedCacheFormat>)primaryFormat
+                         otherCacheFormats:(NSArray<id<MSIDSharedCacheFormat>> *)cacheFormats
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _primaryFormat = primaryFormat;
+        
+        NSMutableArray *allFormatsArray = [@[primaryFormat] mutableCopy];
+        [allFormatsArray addObjectsFromArray:cacheFormats];
+        _allFormats = allFormatsArray;
+    }
+    
+    return self;
+}
+
+#pragma mark - Save tokens
+
+- (BOOL)saveTokensWithRequestParams:(MSIDRequestParameters *)requestParams
+                           response:(MSIDTokenResponse *)response
+                            context:(id<MSIDRequestContext>)context
+                              error:(NSError **)error
+{
+    MSIDAccount *account = [[MSIDAccount alloc] initWithTokenResponse:response];
+    
+    if (response.isMultiResource)
+    {
+        // Save ADAL access token item
+        MSIDToken *accessToken = [[MSIDToken alloc] initWithTokenResponse:response
+                                                                  request:requestParams
+                                                                tokenType:MSIDTokenTypeAccessToken];
+        
+        BOOL result = [_primaryFormat saveAccessToken:accessToken
```

Comment:
_primaryFormat saveAccessToken needed if we already have primary in all Formats?